share/ipld/get.go

```diff
@@ -216,84 +225,131 @@ func GetLeavesByNamespace(
 	// on the level above, the length of the Row is passed in as maxShares
 	leaves := make([]ipld.Node, maxShares)
 
+	// if non-nil proof container provided, fill it while traversing the tree
+	var collectProofs = proofContainer != nil
+	var leftProofs, rightProofs [][]byte
+	if collectProofs {
+		// TODO: (@walldiss) this is massively overallocating and should be optimized later with clever
+		// append
+		leftProofs = make([][]byte, BatchSize(maxShares))
+		rightProofs = make([][]byte, BatchSize(maxShares))
```

Comment:
It is a massive overallocation. `BatchSize` gives the number of nodes in the whole EDS while we work here within one row. You can use `((n*2)-1)-n`, where `(n*2)-1)` is the number of nodes in the tree and `-n` to cut the shares. We can also try to cut these in half more and simplify formula to get left and right proofs exactly.

Reply:
Half the size won't work here, because both arrays represent full tree with exactly `BatchSize` amount of nodes.

Ideally size of each array could be reduced to `treeHeight`. It will require smart append, because nodes needs to be ordered  according to indexes. I wanted to make it as separate PR but can add here as extra commit.

Reply:
Extra commit is fine

Reply:
Done. I also like how GetProofsByNamespace became a bit smaller and simpler now, because I've offload proof collection logic into another simple component.